sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/TransformWithStateExec.scala

```diff
@@ -198,7 +204,37 @@ case class TransformWithStateExec(
       getOutputRow(obj)
     }
     ImplicitGroupingKeyTracker.removeImplicitKey()
-    mappedIterator
+
+    // Wrapper to ensure that the implicit key is set when the methods on the iterator
+    // are called. Inside of processNewData, we use a GroupedIterator, so handleInputRows
+    // is only called once per key. As such, we only have to set the implicit key when
+    // the first call to hasNext is made, and we have to remove it when hasNext returns
+    // false.
+    //
+    // Note: if we ever start to interleave the processing of the iterators we get back
+    // from handleInputRows (i.e. we don't process each iterator all at once), then this
+    // iterator will need to set/unset the implicit key every time hasNext/next is called,
+    // not just at the first and last calls to hasNext.
+    new Iterator[InternalRow] {
+      var hasStarted = false
+
+      override def hasNext: Boolean = {
+        if (!hasStarted) {
+          hasStarted = true
+          ImplicitGroupingKeyTracker.setImplicitKey(keyObj)
+        }
+
+        val hasNext = mappedIterator.hasNext
+        if (!hasNext) {
+          ImplicitGroupingKeyTracker.removeImplicitKey()
+        }
+        hasNext
+      }
+
+      override def next(): InternalRow = {
+        mappedIterator.next()
```

Comment:
I was thinking we also add this block here:

```
        if (!hasStarted) {
          hasStarted = true
          ImplicitGroupingKeyTracker.setImplicitKey(keyObj)
        }

       mappedIterator.next()
```

Reply:
So actually, I was thinking—why would that be needed? We only internally call the methods on this iterator, and we shouldn't do it incorrectly. If anything, we should `assert(hasStarted)` because we should always call `hasNext` before `next`.

Reply:
yea - we could also assert for `hasStarted`